src/sidebar/components/markdown-editor.js

```diff
@@ -134,69 +136,180 @@ ToolbarButton.propTypes = {
   label: propTypes.string,
   onClick: propTypes.func,
   shortcutKey: propTypes.string,
+  tabIndex: propTypes.string,
   title: propTypes.string,
 };
-
+/**
+ * An array of toolbar elements with a roving tab stop. Left and right
+ * array keys can be used to change focus of the elements. Home and end
+ * keys will navigate to the first and last elements respectively.
+ *
+ * Canonical example
+ * https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html
+ */
 function Toolbar({ isPreviewing, onCommand, onTogglePreview }) {
+  // Total number of elements in the toolbar minus 1.
+  const maxElementIndex = 9;
+
+  // Keep track of a roving index. The active roving tabIndex
+  // is set to 0, and all other elements are set to -1.
+  const [rovingElement, setRovingElement] = useState(0);
+
+  // An array of refs. Note we can't use `useRef` in a loop.
+  const elementsRefs = [];
+  for (let i = 0; i <= maxElementIndex; i++) {
+    elementsRefs.push(createRef());
+  }
+
+  /**
+   * Set the element to be focused and set it to be
+   * the active roving index.
+   *
+   * @param {number} index - Ordered index that matches the element
+   */
+  const setFocusedElement = index => {
+    setRovingElement(index);
+    try {
+      // preact component DOM elements are under 'base'
+      elementsRefs[index].current.base.focus();
```

Comment:
One approach to avoid needing the component case would be to pass the ref down to `ToolbarButton` via a different prop (eg. `buttonRef`) which `ToolbarButton` then passes as the `ref` directly to the `<button>`.

```js
<ToolbarButton buttonRef={elementsRefs...}/>

function ToolbarButton({ buttonRef, ... }) {
  return <button ref={buttonRef} .../>;
}
```

This avoids coupling to a Preact-specific API. I've seen some discussion (don't have a link ATM, but can try to find it) that in future in React/Preact it might be possible to access the `ref` prop directly from within the component so that it ends up referring to whatever the component wants it to refer to (eg. the `<button>`) rather than the component itself.

Reply:
Yes, I was going to do that, then forgot! That will indeed be better.